modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java

```diff
@@ -501,6 +543,222 @@ public int getZoneId(String name) {
+
+        return getDataNodesFuture(zoneId, topVerScaleUpFut, topVerScaleDownFut);
+    }
+
+    /**
+     * Waits for DistributionZoneManager observe passed topology version or greater version in topologyWatchListener.
+     *
+     * @param topVer Topology version.
+     * @return Future for chaining.
+     */
+    private CompletableFuture<Void> awaitTopologyVersion(long topVer) {
+        synchronized (dataNodesMutex) {
+            if (topVer > lastTopVer) {
+                return topVerFutures.computeIfAbsent(topVer, key -> new CompletableFuture<>());
+            } else {
+                return completedFuture(null);
+            }
+        }
+    }
+
+    /**
+     * Transforms {@link DistributionZoneConfigurationSchema#dataNodesAutoAdjustScaleUp}
+     * and {@link DistributionZoneConfigurationSchema#dataNodesAutoAdjustScaleDown} values to boolean values.
+     * True if it equals to zero and false if it greater than zero. Zero means that data nodes changing must be scheduled immediate.
+     *
+     * @param zoneId Zone id.
+     * @param topVerFut Future for chaining.
+     * @return Future.
+     */
+    private CompletableFuture<IgniteBiTuple<Boolean, Boolean>> getImmediateTimers(int zoneId, CompletableFuture<Void> topVerFut) {
+        CompletableFuture<IgniteBiTuple<Boolean, Boolean>> timerValuesFut = new CompletableFuture<>();
+
+        topVerFut.thenAccept(ignored -> {
+            DistributionZoneConfiguration zoneCfg = getZoneById(zonesConfiguration, zoneId);
+
+            timerValuesFut.complete(new IgniteBiTuple<>(
+                    zoneCfg.dataNodesAutoAdjustScaleUp().value() == 0,
+                    zoneCfg.dataNodesAutoAdjustScaleDown().value() == 0
+            ));
+        });
+
+        return timerValuesFut;
+    }
+
+    /**
+     * If the {@link DistributionZoneConfigurationSchema#dataNodesAutoAdjustScaleUp} equals to 0 than waits for writing data nodes triggered
+     * by started nodes and corresponding to the passed topology version or greater topology version
+     * to the data nodes into the meta storage.
+     *
+     * @param zoneId Zone id.
+     * @param timerValuesFut Future for chaining.
+     * @return Future for chaining.
+     */
+    private CompletableFuture<Void> scaleUpAwaiting(
+            int zoneId,
+            CompletableFuture<IgniteBiTuple<Boolean, Boolean>> timerValuesFut
+    ) {
+        CompletableFuture<Void> topVerScaleUpFut = new CompletableFuture<>();
+
+        timerValuesFut.thenAccept(timerValues -> {
+            boolean immediateScaleUp = timerValues.get1();
+
+            synchronized (dataNodesMutex) {
+                ZoneState zoneState = zonesState.get(zoneId);
+
+                CompletableFuture<Void> topVerScaleUpFut0 = null;
+
+                if (immediateScaleUp) {
+                    if (zoneState.scaleUpRevision() < lastScaleUpRevision) {
+                        Map.Entry<Long, CompletableFuture<Void>> ceilingEntry =
```

Comment:
Could you please explain what are you doing here? I mean not only the line, but the scaleUpAwaiting itself?
The code seams to be too verbose. I'd expect simpler logic here.
if (zonesState.scaleUpRevision < lastScaleUpRevision){
  await();
}

By the way you have several places of similar logic for awaiting some long values: lastTopVer, lastScaleUpRevision, lastScaleDownRevision.
Instead of explicit future creation I'd rather implement some wrapper class e.g. LongAwaiter or even generic one with CompletableFuture<Long> await(value) method that will handle all that internally and atomically. Maybe we even have such class. It seams to be a sort of long -> long VersiionedValue. 
With such Awaiter it'll be much easier to implement all that lastTopVer, lastScaleDownRevision awaiting logic.

Reply:
Thanks. I fixed it by using PendingComparableValuesTracker.